src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,256 @@
+  // For a table, it holds by contract that either ALL segments are reference segments or ALL segments are not.
+  // For a table, it holds by contract that either ALL segments are reference segments or NONE segments are.
+  // (Mixed reference/not-reference tables don't make sense because we don't want to copy any data if we don't need to).
+  // Therefore, if casting our test_segment to a reference_segment succeeds, we know that we are dealing with a
+  // derived table that is only pointing to a materialized one.
+  return !std::dynamic_pointer_cast<ReferenceSegment>(test_segment);
+}
+
+std::shared_ptr<const std::vector<ChunkOffset>> TableScan::_filter(const std::string& column_type,
+                                                                   std::shared_ptr<const Chunk> chunk) {
+  const auto target_segment = chunk->get_segment(_column_id);
+  const auto segment_size = target_segment->size();
+  auto filtered_positions = std::make_shared<std::vector<ChunkOffset>>();
+  filtered_positions->reserve(segment_size);
+
+  const auto segment_filter_function = _filter_function_for_segment(column_type, target_segment);
+  for (auto row_index = ChunkOffset{0}; row_index < segment_size; ++row_index) {
+    if (segment_filter_function(row_index)) {
+      // Iff the filter function returns true, the row is matched by the filter.
+      filtered_positions->push_back(row_index);
+    }
+  }
+
+  filtered_positions->shrink_to_fit();
+  return filtered_positions;
+}
+
+std::function<bool(ChunkOffset)> TableScan::_filter_function_for_segment(
+    const std::string& column_type, const std::shared_ptr<AbstractSegment>& target_segment) {
+  // In order to avoid having to perform the pointer casting and the switch on the scan type every time we want to
+  // construct a filter function for a segment (which would need to happen on every row of a ReferenceSegment!),
+  // we store our finished filter functions in a map so that we have to construct them only once for each
+  // segment.
+  // Note that we use raw AbstractSegment pointers as the map key, as we want to avoid problems with different
+  // shared_ptr copies to the same segment being treated as different keys. This appears not to be a problem in
+  // practice, but it feels like we shouldn't rely on that.
+  if (!_filter_functions.contains(target_segment.get())) {
+    resolve_data_type(column_type, [this, &column_type, &target_segment](auto type) {
+      using ColumnType = typename decltype(type)::type;
+      Assert(typeid(ColumnType) == _search_value.type(),
+             "Search value doesn't have the same type as the column, we want to compare it with.");
+      const auto _typed_search_value = type_cast<ColumnType>(_search_value);
+
+      std::function<bool(ChunkOffset)> filter_function;
+      if (auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(target_segment); reference_segment) {
```

Comment:
Well, the assignment is performed before actually evaluating the condition (note the `;` on this line) and the condition is essentially the same as `reference_segment != nullptr`. I actually like it this way, because it allows us to perform the assignment close to where the value is actually used (doesn't change much for this branch, but for the other two below). As a nice side effect, we can actually save dynamic pointer casts this way if we enter one of the first two branches because the `else if()` or `else()` would not be evaluated anymore.